File: app/src/main/java/com/example/dieunnph15766_ass/database/IncomeDB.java
package com.example.dieunnph15766_ass.database;

import android.content.ContentValues;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;

import com.example.dieunnph15766_ass.dao.IncomeDao;
import com.example.dieunnph15766_ass.model.Income;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;

public class IncomeDB implements IncomeDao {
    private Database db;
    private SQLiteDatabase sqliteDatabase;
    private ArrayList<Income> list;

    public IncomeDB(Database db) {
        this.db = db;
    }



    @NotNull
    @Override
    public ArrayList<Income> getAllIncomes() {
        sqliteDatabase = db.getWritableDatabase();
        Cursor cursor = sqliteDatabase.rawQuery("SELECT * FROM " + Database.TABLE_INCOME, null);

        if (cursor.getCount() > 0) {
            cursor.moveToFirst();
            while(!cursor.isAfterLast()) {
                Income income = new  Income();
                income.setIncomeID(cursor.getLong(0));
                income.setIncomeName(cursor.getString(1));
                income.setIncomeDate(cursor.getString(2));
                income.setIncomeTypeName(cursor.getString(3));
                income.setUserID(cursor.getLong(4));
                income.setIncomeAmount(cursor.getLong(5));

                list.add(income);
                cursor.moveToNext();
            }
        }

        cursor.close();
        sqliteDatabase.close();

        return list;
    }

    @NotNull
    @Override
    public Income getIncome(@NotNull int index) {
        return list.get(index);
    }

    @Override
    public boolean newIncome(@NotNull Income income) {
        sqliteDatabase = db.getWritableDatabase();
        ContentValues values = new ContentValues();

        values.put("incomeID", income.getIncomeID());
        values.put("incomeName", income.getIncomeName());
        values.put("incomeDate", income.getIncomeDate());
        values.put("incomeTypeName", income.getIncomeTypeName());
        values.put("userID", income.getUserID());
        values.put("incomeAmount", income.getIncomeAmount());

        return sqliteDatabase.insert(Database.TABLE_INCOME, null, values) >0;
    }

    @Override
    public boolean editIncome(@NotNull Income income) {
        sqliteDatabase = db.getWritableDatabase();
        ContentValues values = new ContentValues();

        values.put("incomeID", income.getIncomeID());
        values.put("incomeName", income.getIncomeName());
        values.put("incomeDate", income.getIncomeDate());
        values.put("incomeTypeName", income.getIncomeTypeName());
        values.put("userID", income.getUserID());
        values.put("incomeAmount", income.getIncomeAmount());

        return sqliteDatabase.update(Database.TABLE_INCOME, values, "incomeID = ?", new String[]{""+income.getIncomeID()}) > 0;
    }

    @Override
    public boolean removeIncome(@NotNull Income income) {
        sqliteDatabase = db.getWritableDatabase();
        return sqliteDatabase.delete(Database.TABLE_INCOME, "incomeID = ?", new String[]{"" + income.getIncomeID()}) > 0;
    }
}

File: app/src/main/java/com/example/dieunnph15766_ass/database/income/IncomeDB.kt
package com.example.dieunnph15766_ass.database.income

import android.content.ContentValues
import android.database.sqlite.SQLiteDatabase
import com.example.dieunnph15766_ass.dao.income.IncomeDao
import com.example.dieunnph15766_ass.database.Database
import com.example.dieunnph15766_ass.model.income.Income
import java.util.*

class IncomeDB(private val db: Database) : IncomeDao {
    private var sqliteDatabase: SQLiteDatabase? = null
    private var list: ArrayList<Income>? = null

    override fun getAllIncomes(username: String): ArrayList<Income> {
        sqliteDatabase = db.writableDatabase
        val cursor = sqliteDatabase!!.rawQuery("SELECT * FROM ${Database.TABLE_INCOME} WHERE USERNAME = \"${username}\" " , null)
        list = ArrayList()
        if (cursor.count > 0) {
            cursor.moveToFirst()
            while (!cursor.isAfterLast) {
                val income = Income()
                income.incomeID = cursor.getLong(0)
                income.incomeName = cursor.getString(1)
                income.incomeDate = cursor.getString(2)
                income.incomeTypeName = cursor.getString(3)
                income.userName = cursor.getString(4)
                income.incomeAmount = cursor.getLong(5)
                income.incomeNote = cursor.getString(6)
                list!!.add(income)
                cursor.moveToNext()
            }
        }
        cursor.close()
        sqliteDatabase!!.close()
        return list!!
    }

    override fun getIncome(index: Int): Income {
        return list!![index]
    }

    override fun newIncome(income: Income): Boolean {
        sqliteDatabase = db.writableDatabase
        val values = ContentValues()
        values.put("INCOME_ID", income.incomeID)
        values.put("INCOME_NAME", income.incomeName)
        values.put("INCOME_DATE", income.incomeDate)
        values.put("INCOME_TYPE_NAME", income.incomeTypeName)
        values.put("USERNAME", income.userName)
        values.put("INCOME_AMOUNT", income.incomeAmount)
        values.put("INCOME_NOTE", income.incomeNote)
        return sqliteDatabase!!.insert(Database.TABLE_INCOME, null, values) > 0
    }

    override fun editIncome(income: Income, username: String): Boolean {
        sqliteDatabase = db.writableDatabase
        val values = ContentValues()
        values.put("INCOME_NAME", income.incomeName)
        values.put("INCOME_DATE", income.incomeDate)
        values.put("INCOME_TYPE_NAME", income.incomeTypeName)
        values.put("USERNAME", income.userName)
        values.put("INCOME_AMOUNT", income.incomeAmount)
        values.put("INCOME_NOTE", income.incomeNote)
        return sqliteDatabase!!.update(
            Database.TABLE_INCOME,
            values,
            "INCOME_NAME = ? AND USERNAME = ?",
            arrayOf(income.incomeName, username)
        ) > 0
    }

    override fun removeIncome(income: Income, username: String): Boolean {
        sqliteDatabase = db.writableDatabase
        return sqliteDatabase!!.delete(
            Database.TABLE_INCOME,
            "INCOME_NAME = ? AND USERNAME = ?",
            arrayOf(income.incomeName, username)
        ) > 0
    }
}